File: 1200-1400/1249. Minimum Remove to Make Valid Parentheses/solution.js
var minRemoveToMakeValid = function(s) {
    let str = '', leftCount = 0
    for (let i=0; i<s.length; i++) {
        if (/[a-z]/.test(s[i])) {
            str += s[i]
        } else {
            if(s[i] === '(') {
                str += s[i]
                leftCount++
            } else if (leftCount>0) {
                str += s[i]
                leftCount--
            }
        }
    }
    let rightCount = 0, ans = ''
    for (let i=str.length-1; i>=0; i--) {
        if (/[a-z]/.test(str[i])) {
            ans = str[i] + ans
        } else {
            if (str[i] === ')') {
                ans = str[i] + ans
                rightCount++
            } else if (rightCount>0) {
                ans = str[i] + ans
                rightCount--
            }
        }
    }
  return ans  
};

File: 1200-1400/1249. Minimum Remove to Make Valid Parentheses/solution.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        vector<int> left;
        vector<int> right;
        for (int i=0; i<s.size(); i++) {
            if (s[i] == ')') {
                if (left.empty()) right.push_back(i);
                else left.pop_back();
            } else if (s[i] == '(') {
                left.push_back(i);
            }
        }
        
        string ans = "";
        for (int i=0, j=0, k=0; i<s.size(); i++) {
            if (j<right.size() && i == right[j]) j++;
            else if (k<left.size() && i == left[k]) k++;
            else ans += s[i];
        }
        return ans;
    }
};


/*
valid => pairs of paren
check if ( == )
if ( == ) cool
if ( > ) skip the first ( - ) '('
if ( < ) skip the 

every ) is valid if there is a vacant (  before, vice versas

1. record the indexes of (
2. if met with (  than pop out
3. also store indices of lonely )

*/
